src/main/java/com/redhat/lightblue/hook/publish/IdentityExctractionUtil.java

```diff
@@ -0,0 +1,191 @@
+        }
+
+        if (!(preObject == null || preObject.getClass().equals(postObject.getClass())) || !postObject.getClass().equals(idsObject.getClass())) {
+            throw new IllegalArgumentException("pre state (optional), post state and projected ids need to be valid JSON of the same type:: "
+                    + getErrorSignature(pre, post, ids));
+        }
+        // JSONParser only returns JSONArray/ JSONObject/ JSONString
+        if (!postObject.getClass().equals(JSONArray.class) && !postObject.getClass().equals(JSONObject.class)) {
+            throw new IllegalArgumentException("Identities can only extracted from JSONArrays or JSONObjects:: " + getErrorSignature(pre, post, ids));
+        }
+
+        return compareAndGetIdentities(preObject, postObject, idsObject);
+    }
+
+    /*
+     * For child objects and child arrays, cross multiply the number of simple
+     * fields. This is because we need one event per child array element that
+     * changed.
+     */
+    public static Set<IdentitySet> compareAndGetIdentities(JSONObject pre, JSONObject post, JSONObject ids) throws JSONException {
+        Set<IdentitySet> result = new HashSet<>();
+        Iterator<String> keysIterator = ids.keys();
+        while (keysIterator.hasNext()) {
+            String key = keysIterator.next();
+            if (post.get(key) instanceof JSONObject || post.get(key) instanceof JSONArray) {
+                Object preObject = (pre == null) ? null : pre.opt(key);
+                Set<IdentitySet> arraySets = getNewSetsWithPrependName(key, compareAndGetIdentities(preObject, post.get(key), ids.get(key)));
+                result = crossMultiplySets(result, arraySets);
+            } else {
+                Identity i = new Identity();
+                i.setField(key);
+                i.setValue(post.get(key).toString());
+                String operation = (pre == null) ? INSERT_OPERATION : UPDATE_OPERATION;
+                if (result.size() == 0) {
+                    result.add(new IdentitySet(new HashSet<Identity>(), operation));
+                }
+                for (IdentitySet set : result) {
+                    set.getSet().add(i);
+                }
+            }
+        }
+        return result;
+    }
+
+    /*
+     * we cannot assume order of objects in the array are going to remain the
+     * same, however, we can assume the objects in the array are unique in they
+     * identifying field values.
+     * 
+     * identity object is the subset of the integrated state of the entity.
+     * using that property we create a map for both pre and post arrays, with
+     * the identifying sub state as the key.
+     * 
+     * then for all array elements in the identifying array we lookup if that
+     * element was added or updated.
+     */
+    public static Set<IdentitySet> compareAndGetIdentities(JSONArray pre, JSONArray post, JSONArray ids) throws JSONException {
```

Comment:
@vritant I was looking at this purely from readability standpoint. Just curious what your reasons were.